url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8/src/com/fuzzylite/imex/FllExporter.java

```diff
@@ -0,0 +1,173 @@
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+package com.fuzzylite.imex;
+
+import com.fuzzylite.Engine;
+import com.fuzzylite.Op;
+import com.fuzzylite.defuzzifier.Defuzzifier;
+import com.fuzzylite.defuzzifier.IntegralDefuzzifier;
+import com.fuzzylite.norm.Norm;
+import com.fuzzylite.rule.Rule;
+import com.fuzzylite.rule.RuleBlock;
+import com.fuzzylite.term.Term;
+import com.fuzzylite.variable.InputVariable;
+import com.fuzzylite.variable.OutputVariable;
+import com.fuzzylite.variable.Variable;
+
+/**
+ *
+ * @author jcrada
+ */
+public class FllExporter extends Exporter {
+
+    protected String indent;
+    protected String separator;
+
+    public FllExporter() {
+        this("  ", "\n");
+    }
+
+    public FllExporter(String indent, String separator) {
+        this.indent = indent;
+        this.separator = separator;
+    }
+
+    public String getIndent() {
+        return indent;
+    }
+
+    public void setIndent(String indent) {
+        this.indent = indent;
+    }
+
+    public String getSeparator() {
+        return separator;
+    }
+
+    public void setSeparator(String separator) {
+        this.separator = separator;
+    }
+
+    @Override
+    public String toString(Engine engine) {
+        StringBuilder result = new StringBuilder();
+        result.append(String.format("Engine: %s%s", engine.getName(), separator));
+        for (InputVariable inputVariable : engine.getInputVariables()) {
+            result.append(toString(inputVariable));
+        }
+        for (OutputVariable outputVariable : engine.getOutputVariables()) {
+            result.append(toString(outputVariable));
+        }
+        for (RuleBlock ruleBlock : engine.getRuleBlocks()) {
+            result.append(toString(ruleBlock));
+        }
+        return result.toString();
+    }
+
+    public String toString(Variable variable) {
+        StringBuilder result = new StringBuilder();
+        result.append(String.format("Variable: %s%s", variable.getName(), separator));
+        result.append(String.format("%senabled: %s%s", indent,
+                String.valueOf(variable.isEnabled()), separator));
+        result.append(String.format("%srange: %s%s", indent,
+                Op.join(" ", variable.getMinimum(), variable.getMaximum()), separator));
+        for (Term term : variable.getTerms()) {
+            result.append(String.format("%s%s%s", indent, toString(term), separator));
+        }
+        return result.toString();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 80:89, 93:102.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.